### PR TITLE
Ret 1455

### PR DESCRIPTION
**Pull request by robin-ieong**
https://tools.hmcts.net/jira/browse/RET-1455

### Change description ###
Add API integration and functional tests including a functional health check test.

 

Tasks:

Integration tests required to mock calls to external services like CCD APIs
Functional tests - make sure that API end point is working end to end with a functional test (no mocking)
Functional tests required for email notifications via Gov.UK Notify service
Create functional health check endpoint that can be called from Jenkins pipelines.
 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/et/syaapi/CreateUser.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CreateUser {
```
Comment (nickwest1) [inline src/functionalTest/java/uk/gov/hmcts/reform/et/syaapi/CreateUser.java:11]: Can't we use @Data, @AllArgsConstructor and @Getter here?

File: src/functionalTest/java/uk/gov/hmcts/reform/et/syaapi/IdamTestApiRequests.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        assertEquals(OK.value(), loginResponse.getStatusLine().getStatusCode());
+
+        String tokens = EntityUtils.toString(loginResponse.getEntity());
+        JSONObject jsonObject;
+        String accessToken = null;
+        try {
+            jsonObject = new JSONObject(tokens);
+            accessToken = jsonObject.get("access_token").toString();
+        } catch (JSONException e) {
+            log.error("Failed to get access token from loginResponse");
```
Comment (nickwest1) [inline src/functionalTest/java/uk/gov/hmcts/reform/et/syaapi/IdamTestApiRequests.java:82]: Include the exception on logging:
log.error("Failed to get access token from loginResponse", e);

Beyond this, if this fails, then should you not be throwing an exception on up given you can't extract the access token?  Don't just ignore that.

File: src/functionalTest/java/uk/gov/hmcts/reform/et/syaapi/Role.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.hmcts.reform.et.syaapi;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Role {
```
Comment (nickwest1) [inline src/functionalTest/java/uk/gov/hmcts/reform/et/syaapi/Role.java:9]: Again, should be able to use @DaTa, @AllArgsConstructor and @getter here, surely?

File: src/integrationTest/java/uk/gov/hmcts/reform/et/syaapi/controllers/ManageCaseControllerIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    @MockBean
+    private CoreCaseDataApi ccdApiClient;
+
+    public ManageCaseControllerIntegrationTest() throws IOException {
+        // Just to avoid multiple try-catch
+    }
+
+    @BeforeEach
+    void setUp() {
+        mockMvc = webAppContextSetup(wac).build();
```
Comment (jthmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/et/syaapi/controllers/ManageCaseControllerIntegrationTest.java:76]: You've already autowired this

File: src/integrationTest/java/uk/gov/hmcts/reform/et/syaapi/controllers/ResourceLoader.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    private static String resourceAsString(final String resourcePath) throws IOException {
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        final File file = ResourceUtils.getFile(classLoader.getResource(resourcePath).getFile());
+        return new String(Files.readAllBytes(file.toPath()));
+    }
+
+}
```
Comment (nickwest1) [inline src/integrationTest/java/uk/gov/hmcts/reform/et/syaapi/controllers/ResourceLoader.java:52]: Prefer not a static class, then you can inject the ObjectMappaer already available as a bean

**Review by nickwest1 — CHANGES_REQUESTED** [review]

**Review by nickwest1 — APPROVED** [review]

**Review by jthmcts — APPROVED** [review]

**Pull request merged**